src/elementary/ParticleID.hpp

```diff
@@ -0,0 +1,107 @@
+#ifndef NJOY_ELEMENTARY_PARTICLEID
+#define NJOY_ELEMENTARY_PARTICLEID
+
+// system includes
+#include <string>
+#include <variant>
+
+// other includes
+#include "elementary/src/overload.hpp"
+#include "elementary/FundamentalParticleID.hpp"
+#include "elementary/NucleusID.hpp"
+#include "elementary/NuclideID.hpp"
+
+namespace njoy {
+namespace elementary {
+
+  /**
+   *  @class
+   *  @brief The particle identifier, either a fundamental particle, a nucleus
+   *         or nuclide identifier
+   */
+  class ParticleID {
+
+    /* type aliases */
+    using IdentifierVariant = std::variant< FundamentalParticleID, NucleusID, NuclideID >;
+
+    /* fields */
+    IdentifierVariant identifier_;
+
+    /* auxiliary functions */
+
+  public:
+
+    /* constructor */
+    #include "elementary/ParticleID/src/ctor.hpp"
+
+    /* methods */
+
+    /**
+     *  @brief return the particle symbol
+     */
+    std::string symbol() const noexcept {
+
+      return std::visit( [] ( const auto& identifier )
+                            { return identifier.symbol(); },
+                         this->identifier_ );
+    }
+
+    /**
+     *  @brief return the nuclide name
+     */
+    std::string name() const noexcept {
+
+      return std::visit( [] ( const auto& identifier )
+                            { return identifier.name(); },
+                         this->identifier_ );
+    }
+
+    /**
+     *  @brief return the nuclide hash value
+     */
+    int hash() const noexcept {
+
+      return std::visit(
+               overload{ [] ( const FundamentalParticleID& identifier )
+                            { return identifier.hash(); },
+                         [] ( const NucleusID& identifier )
+                            { return identifier.hash() * 10; },
+                         [] ( const NuclideID& identifier )
+                            { return identifier.hash() * 10 + 1; } },

```

Comment:
This is cool. Didn't know you could do that. Where did you learn about this?

Reply:
It's called google ...

Reply:
I had already heard of overloading lambdas, and it makes for nice code so I used it.

Reply:
I googled "overload lambda c++" and found something, but wouldn't have known to even try had I not seen this.